subsys/net/lib/nrf_cloud/src/nrf_cloud_fota.c

```diff
@@ -0,0 +1,1251 @@
+
+static int fota_settings_set(const char *key, size_t len_rd,
+			     settings_read_cb read_cb, void *cb_arg)
+{
+	if (!key) {
+		LOG_DBG("Key is NULL");
+		return -EINVAL;
+	}
+
+	LOG_DBG("Settings key: %s, size: %d", log_strdup(key), len_rd);
+
+	if (!strncmp(key, SETTINGS_FOTA_JOB, strlen(SETTINGS_FOTA_JOB)) &&
+	    (len_rd == sizeof(saved_job))) {
+		if (read_cb(cb_arg, (void *)&saved_job, len_rd) == len_rd) {
+			LOG_DBG("Saved job: %s, type: %d, validate: %d",
+				log_strdup(saved_job.id), saved_job.type,
+				saved_job.validate);
+			return 0;
+		}
+	}
+	return -ENOTSUP;
+}
+
+static enum fota_validate_status get_modem_update_status(void)
+{
+	enum fota_validate_status ret = NRF_CLOUD_FOTA_VALIDATE_UNKNOWN;
+
+#if defined(CONFIG_NRF_MODEM_LIB)
+	int modem_dfu_res = nrf_modem_lib_get_init_ret();
+
+	/* Handle return values relating to modem firmware update */
+	switch (modem_dfu_res) {
+	case MODEM_DFU_RESULT_OK:
+		LOG_DBG("Modem FOTA OK");
+		ret = NRF_CLOUD_FOTA_VALIDATE_PASS;
+		break;
+	case MODEM_DFU_RESULT_UUID_ERROR:
+	case MODEM_DFU_RESULT_AUTH_ERROR:
+	case MODEM_DFU_RESULT_HARDWARE_ERROR:
+	case MODEM_DFU_RESULT_INTERNAL_ERROR:
+		LOG_ERR("Modem FOTA error: %d", modem_dfu_res);
+		ret = NRF_CLOUD_FOTA_VALIDATE_FAIL;
+		break;
+	default:
+		LOG_DBG("Modem FOTA result unknown: %d", modem_dfu_res);
+		break;
+	}
+#endif /* CONFIG_NRF_MODEM_LIB */
+
+	return ret;
+}
+
+int nrf_cloud_fota_init(nrf_cloud_fota_callback_t cb)
+{
+	int ret;
+	bool update_was_pending = false;
+	enum fota_validate_status validate = NRF_CLOUD_FOTA_VALIDATE_UNKNOWN;
+
+	if (cb == NULL) {
+		LOG_ERR("Invalid parameter");
+		return -EINVAL;
+	}
+
+	event_cb = cb;
+
+	if (initialized) {
+		return 0;
+	}
+
+	ret = fota_download_init(http_fota_handler);
+	if (ret != 0) {
+		LOG_ERR("fota_download_init error: %d", ret);
+		return ret;
+	}
+
+	ret = settings_load_subtree(settings_handler_fota.name);
+	if (ret) {
+		LOG_ERR("Cannot load settings: %d", ret);
+		return ret;
+	}
+
+	if (saved_job.validate == NRF_CLOUD_FOTA_VALIDATE_PENDING) {
+		update_was_pending = true;
+#if defined(CONFIG_BOOTLOADER_MCUBOOT)
+		if (!boot_is_img_confirmed()) {
+			ret = boot_write_img_confirmed();
+			if (ret) {
+				LOG_ERR("FOTA update confirmation failed: %d",
+					ret);
+				/* If this fails then MCUBOOT will revert
+				 * to the previous image on reboot
+				 */
+				validate = NRF_CLOUD_FOTA_VALIDATE_FAIL;
+				update_was_pending = false;
+			} else {
+				LOG_DBG("FOTA update confirmed");
+				validate = NRF_CLOUD_FOTA_VALIDATE_PASS;
+			}
+		}
+#endif
+
+		if (saved_job.type == NRF_CLOUD_FOTA_MODEM) {
+			validate = get_modem_update_status();
+		}
+
+		/* save status and update when cloud connection is ready */
+		save_validate_status(saved_job.id, saved_job.type, validate);
+
+		if (saved_job.type == NRF_CLOUD_FOTA_MODEM) {
+			/* Reboot is required if type is still set to MODEM */
+			LOG_INF("Rebooting to complete modem FOTA");
+			sys_reboot(SYS_REBOOT_COLD);
```

Comment:
Is it possible to propagate an event here to the app instead of rebooting in the library? Then the application can have the choice to reboot or just restart the modem and not the application

Reply:
yes, i will create a JIRA task to improve this.  
i still wasn't able to get the modem-restart-without-reboot working, so i'll try some more for that too.

for modem fota there are 2 reboots that occur.  the initial reboot, which allows for the modem to install.
then, this, the reboot after the successful installation.

Reply:
Oh okay, I thought it was just the one reboot. I haven't tested the "warm reboot" of the modem myself, but sounds like a useful feature in these cases. But we might run into issues with stuff that has initialized and set up sockets during boot, that will probably fail if modem is rebooted but their init functions are not run again. For instance common AT interface and GPS driver.

Reply:
CIA-175 for reference